Migration: Add auto-automigration for merged packages

ndelangen opened the pull request:
## What I did

Added an automigration for CPC compliance.
No code should use the following packages anymore:
- `@storybook/channels`
- `@storybook/client-logger`
- `@storybook/core-common`
- `@storybook/core-events`
- `@storybook/csf-tools`
- `@storybook/docs-tools`
- `@storybook/node-logger`
- `@storybook/preview-api`
- `@storybook/router`
- `@storybook/telemetry`
- `@storybook/theming`
- `@storybook/types`
- `@storybook/manager-api`
- `@storybook/manager`
- `@storybook/preview`
- `@storybook/core-server`
- `@storybook/builder-manager`
- `@storybook/components`
- `@storybook/test`

These have been merged into `storybook`-package, with these paths:

- `storybook/internal/channels`
- `storybook/internal/client-logger`
- `storybook/internal/common`
- `storybook/internal/core-events`
- `storybook/internal/csf-tools`
- `storybook/internal/docs-tools`
- `storybook/internal/node-logger`
- `storybook/internal/preview-api`
- `storybook/internal/router`
- `storybook/internal/telemetry`
- `storybook/internal/theming`
- `storybook/internal/types`
- `storybook/internal/manager-api`
- `storybook/internal/manager`
- `storybook/internal/preview`
- `storybook/internal/core-server`
- `storybook/internal/builder-manager`
- `storybook/internal/components`
- `storybook/test`

This automigration will:

- [x] search for the user's files, and replace imports and require statements from old to new
- [x] uninstall old packages

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

<!-- Please include the steps to test your changes here. For example:

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Open Storybook in your browser
3. Access X story

-->

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->
This pull request has been released as version `0.0.0-pr-30753-sha-0519e608`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-30753-sha-0519e608 sandbox` or in an existing project with `npx storybook@0.0.0-pr-30753-sha-0519e608 upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-30753-sha-0519e608`](https://npmjs.com/package/storybook/v/0.0.0-pr-30753-sha-0519e608) |
| **Triggered by** | @valentinpalkovic |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`norbert/automigrate-cpc-complaince`](https://github.com/storybookjs/storybook/tree/norbert/automigrate-cpc-complaince) |
| **Commit** | [`0519e608`](https://github.com/storybookjs/storybook/commit/0519e6083ef93fd49f0eccfc4e8adea79b581461) |
| **Datetime** | Sun Mar  9 19:10:52 UTC 2025 (`1741547452`) |
| **Workflow run** | [13751647540](https://github.com/storybookjs/storybook/actions/runs/13751647540) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=30753`_
</details>
<!-- CANARY_RELEASE_SECTION -->

<!-- BENCHMARK_SECTION -->

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createSize |  0 B | 0 B | 0 B | - | - |
| generateSize |  79.9 MB | 80 MB | 59.2 kB | **14.06** | 0.1% |
| initSize |  79.9 MB | 80 MB | 59.2 kB | **14.06** | 0.1% |
| diffSize |  97 B | 97 B | 0 B | - | 0% |
| buildSize |  7.47 MB | 7.47 MB | 0 B | -0.6 | 0% |
| buildSbAddonsSize |  1.98 MB | 1.98 MB | 0 B | 0.9 | 0% |
| buildSbCommonSize |  195 kB | 195 kB | 0 B | - | 0% |
| buildSbManagerSize |  1.88 MB | 1.88 MB | 0 B | - | 0% |
| buildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| buildStaticSize |  0 B | 0 B | 0 B | - | - |
| buildPrebuildSize |  4.06 MB | 4.06 MB | 0 B | 0.9 | 0% |
| buildPreviewSize |  3.42 MB | 3.42 MB | 0 B | **-2.89** | 0% |
| testBuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbAddonsSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbCommonSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbManagerSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| testBuildStaticSize |  0 B | 0 B | 0 B | - | - |
| testBuildPrebuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildPreviewSize |  0 B | 0 B | 0 B | - | - |



| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createTime |  11.5s | 25.4s | 13.8s | 1.14 | 54.6% |
| generateTime |  20.2s | 19.7s | -502ms | 0.15 | -2.5% |
| initTime |  4.9s | 4.7s | -249ms | -0.02 | -5.3% |
| buildTime |  9.1s | 9s | -122ms | -0.37 | -1.3% |
| testBuildTime |  0ms | 0ms | 0ms | - | - |
| devPreviewResponsive |  4.6s | 4.9s | 294ms | -0.33 | 6% |
| devManagerResponsive |  4.4s | 4.7s | 282ms | -0.31 | 6% |
| devManagerHeaderVisible |  782ms | 780ms | -2ms | -0.28 | -0.3% |
| devManagerIndexVisible |  793ms | 794ms | 1ms | -0.41 | 0.1% |
| devStoryVisibleUncached |  2.7s | 2.3s | -345ms | -0.83 | -14.5% |
| devStoryVisible |  1.2s | 1.2s | 31ms | -0.2 | 2.5% |
| devAutodocsVisible |  1.1s | 1.2s | 108ms | 0.48 | 8.8% |
| devMDXVisible |  1.1s | 1s | -59ms | -0.72 | -5.5% |
| buildManagerHeaderVisible |  880ms | 919ms | 39ms | 0.46 | 4.2% |
| buildManagerIndexVisible |  885ms | 964ms | 79ms | 0.68 | 8.2% |
| buildStoryVisible |  869ms | 911ms | 42ms | 0.48 | 4.6% |
| buildAutodocsVisible |  618ms | 689ms | 71ms | -0.76 | 10.3% |
| buildMDXVisible |  666ms | 677ms | 11ms | -0.77 | 1.6% |

<!-- BENCHMARK_SECTION -->


<!-- greptile_comment -->

## Greptile Summary

Implements an automigration to consolidate multiple Storybook packages into the main `storybook` package, updating import paths and removing deprecated package dependencies for CPC compliance.

- Added `code/lib/cli-storybook/src/automigrate/fixes/consolidated-imports.ts` to handle package.json updates and import path transformations
- Added `code/lib/cli-storybook/src/automigrate/helpers/consolidated-packages.ts` defining mappings for 19 packages to their new internal paths
- Added comprehensive test suite in `consolidated-imports.test.ts` covering package detection, transformation, and error handling
- Modified `code/lib/cli-storybook/src/automigrate/fixes/index.ts` to include the new `consolidatedImports` fix



<!-- /greptile_comment -->

ndelangen committed:
Refactor consolidated-imports fix with improved error handling and Babel transformation

ndelangen committed:
improve automigration

ndelangen committed:
Merge branch 'next' into norbert/automigrate-cpc-complaince

ndelangen committed:
Fix: Dynamically import p-limit in consolidated imports

ndelangen committed:
Fix: Correct p-limit import syntax in consolidated imports

ndelangen committed:
Refactor: Extract consolidated packages into a separate helper module

ndelangen committed:
Feature: Add consolidated package.json automigration fix

This commit introduces a new automigration fix to handle consolidated Storybook packages in package.json files. The implementation includes:

- A check to identify package.json files with consolidated packages
- Transformation logic to remove consolidated packages
- Adding the main Storybook package to devDependencies
- Comprehensive test coverage for the new feature

valentinpalkovic commented on code/lib/cli-storybook/src/automigrate/fixes/index.ts:
```diff
@@ -69,6 +70,7 @@ export const allFixes: Fix[] = [
   autodocsTags,
   initialGlobals,
   addonA11yAddonTest,
+  consolidatedImports,
```
`consolidatedPackageJson` is missing here

valentinpalkovic commented on code/lib/cli-storybook/src/automigrate/fixes/consolidated-package-json.ts:
```diff
@@ -0,0 +1,135 @@
+
+export const consolidatedPackageJson: Fix<ConsolidatedPackageJsonOptions> = {
+  id: 'consolidated-package-json',
+  versionRange: ['<8.0.0', '>=8.0.0'],
```
```suggestion
  versionRange: ['<9.0.0', '>=9.0.0'],
```

ndelangen commented on code/lib/cli-storybook/src/automigrate/fixes/consolidated-package-json.ts:
```diff
@@ -0,0 +1,135 @@
+
+    const projectRoot = getProjectRoot();
+    const patterns = ['**/package.json'];
+    const files = (await globby(patterns, {
+      ...commonGlobOptions(''),
+      ignore: ['**/node_modules/**'],
+      cwd: projectRoot,
+    })) as string[];
+
+    // Check if any package.json files contain consolidated packages
+    const filesWithConsolidatedPackages: string[] = [];
+
+    for (const file of files) {
+      const content = await readFile(file, 'utf-8');
+      const packageJson = JSON.parse(content);
+      const hasConsolidatedPackage = Object.keys(consolidatedPackages).some((pkg) => {
+        return (
+          (packageJson.dependencies && pkg in packageJson.dependencies) ||
+          (packageJson.devDependencies && pkg in packageJson.devDependencies)
+        );
+      });
+
+      if (hasConsolidatedPackage) {
+        filesWithConsolidatedPackages.push(file);
+      }
+    }
```
Yeah, I did this to better support monorepos, where I do not quite know where users might have storybook dependencies in which package.json files.

ndelangen committed:
Refactor: Remove unused import in consolidated package.json automigration

ndelangen committed:
Automigrate: Add consolidated package.json fix to allFixes array

ndelangen committed:
Merge branch 'next' into norbert/automigrate-cpc-complaince

ndelangen committed:
Refactor: Consolidate import and package.json migration fixes

This commit merges and refactors the consolidated imports and package.json migration fixes into a single, more comprehensive automigration solution. Key changes include:

- Combine separate import and package.json transformation logic
- Improve error handling and file processing
- Update type definitions and test coverage
- Simplify the overall migration approach for consolidated Storybook packages

ndelangen commented on code/lib/cli-storybook/src/automigrate/fixes/consolidated-imports.ts:
```diff
@@ -0,0 +1,133 @@
+      if (hasConsolidatedImport) {
+        filesWithConsolidatedImports.push(file);
+      }
+    }
```
okay, reasonable. I merged both migrations into a single one.

ndelangen committed:
Merge branch 'next' into norbert/automigrate-cpc-complaince

valentinpalkovic commented on code/lib/cli-storybook/src/automigrate/fixes/consolidated-imports.ts:
```diff
@@ -0,0 +1,215 @@
+    const packageJsonFiles = await globby(['**/package.json'], {
+      ...commonGlobOptions(''),
+      ignore: ['**/node_modules/**'],
+      cwd: projectRoot,
```
```suggestion
      cwd: projectRoot,
      gitignore: true,
```

valentinpalkovic commented on code/lib/cli-storybook/src/automigrate/fixes/consolidated-imports.ts:
```diff
@@ -0,0 +1,215 @@
+  if (!packageJson.devDependencies) {
+    packageJson.devDependencies = {};
+  }
+  if (!packageJson.devDependencies.storybook) {
+    packageJson.devDependencies.storybook = '^8.0.0';
+    hasChanges = true;
+  }
```
In Storybook 8, `storybook` was already required to be present in the user's package.json. It's not necessary to check for it. Additionally, it can also add `storybook` unnecessarily in nested package.jsons, especially in mono repos where NX's single version policy is applied, only the root package.json contains dependencies, whereas leaf package.jsons only defines scripts.

valentinpalkovic commented on code/lib/cli-storybook/src/automigrate/fixes/consolidated-imports.ts:
```diff
@@ -0,0 +1,215 @@
+    parser: {
+      parse(code: string) {
+        return parser.parse(code, {
+          sourceType: 'unambiguous',
+          plugins: ['typescript', 'jsx'],
+        });
+      },
+    },
+  });
+  let hasChanges = false;
+
+  traverse(ast, {
+    ImportDeclaration(path: NodePath<t.ImportDeclaration>) {
+      const importSource = path.node.source.value;
+      if (typeof importSource === 'string' && importSource in consolidatedPackages) {
+        path.node.source = t.stringLiteral(
+          consolidatedPackages[importSource as ConsolidatedPackage]
+        );
+        hasChanges = true;
+      }
+    },
+    CallExpression(path: NodePath<t.CallExpression>) {
+      if (
+        t.isIdentifier(path.node.callee, { name: 'require' }) &&
+        path.node.arguments.length > 0 &&
+        t.isStringLiteral(path.node.arguments[0])
+      ) {
+        const arg = path.node.arguments[0];
+        if (arg.value in consolidatedPackages) {
+          path.node.arguments[0] = t.stringLiteral(
+            consolidatedPackages[arg.value as ConsolidatedPackage]
+          );
+          hasChanges = true;
+        }
+      }
+    },
+  });
+
+  return hasChanges ? recast.print(ast).code : null;
+}
```
You can keep this code and fine-tune to also be able to process `*.stories.svelte* files, but I think this is way too complicated and doesn't support. 

I personally would suggest a dead-simple search and replace as I did here: https://github.com/storybookjs/storybook/pull/30761/files#diff-51aaa23bbf3198e1923160bbc9804d6dd1efc13820673cb2d9314828cef5a5b0R81

valentinpalkovic commented on code/lib/cli-storybook/src/automigrate/fixes/consolidated-imports.ts:
```diff
@@ -0,0 +1,215 @@
+      
+      Would you like to:
+      1. Update these package.json files
+      2. Scan your codebase and update any imports from these consolidated packages
```
Historically, these kind of file transformations were "migrations". Now we want to have them as "automigrations". When used as "migrations", users had the possibility to apply a different glob pattern for the file transformations. I followed this pattern in my PR: https://github.com/storybookjs/storybook/pull/30761 and I think it makes sense to also follow it here (prompting the user for a glob pattern before continuing)

ndelangen committed:
Remove consolidated package.json migration fix

Clean up automigrate fixes by removing the consolidated package.json migration, which appears to have been consolidated into another fix in a previous refactoring.

ndelangen committed:
Automigrate: Add gitignore support to file globbing

Enhance the consolidated imports fix by adding gitignore support to file globbing options, which helps exclude files specified in .gitignore during migration processing.

ndelangen committed:
Automigrate: Remove Storybook version addition to devDependencies

Remove automatic addition of Storybook version to package.json devDependencies in the consolidated imports migration fix. This change simplifies the package.json transformation logic and allows more flexibility in dependency management.

ndelangen committed:
Refactor: Simplify consolidated imports transformation

Streamline the import migration logic by replacing the AST-based transformation with a simpler regex-based approach. This change:
- Reduces complexity of import and require statement replacements
- Uses direct string replacement instead of abstract syntax tree manipulation
- Maintains the core functionality of package import consolidation

ndelangen committed:
Automigrate: Add custom glob pattern support for consolidated imports

Enhance the consolidated imports migration by:
- Introducing an interactive prompt to allow custom file scanning patterns
- Dynamically importing globby to reduce initial import overhead
- Providing a default glob pattern with option for user customization

ndelangen committed:
Automigrate: Add dependency installation for consolidated imports

Enhance the consolidated imports migration by:
- Adding automatic dependency installation after file transformations
- Conditionally running package manager install only when not in dry run mode
- Suppressing an eslint rule for error handling to improve migration robustness

ndelangen committed:
Automigrate: Add '@storybook/test' to consolidated packages

Add the '@storybook/test' package to the consolidated packages mapping, extending the import consolidation support for Storybook's test utilities.

ndelangen committed:
Automigrate: Add documentation for consolidated packages mapping

Add a descriptive comment explaining the purpose of the consolidated packages object, which maps old Storybook package names to their new internal package names.

valentinpalkovic commented on code/lib/cli-storybook/src/automigrate/fixes/consolidated-imports.ts:
```diff
@@ -0,0 +1,199 @@
+
+export const consolidatedImports: Fix<ConsolidatedOptions> = {
+  id: 'consolidated-imports',
+  versionRange: ['^8.0.0', '^9.0.0'],
```
```suggestion
  versionRange: ['^8.0.0', '^9.0.0-0'],
```

To support beta releases, otherwise we will have a hard time to test this.

ndelangen committed:
Automigrate: Dynamically import p-limit for consolidated imports

Update the consolidated imports transformation to use dynamic import for p-limit, reducing initial import overhead and improving module loading flexibility.

ndelangen committed:
Automigrate: Remove dependencies field from BaseFix type

Simplify the BaseFix type by removing the optional dependencies field, which appears to be no longer used in the automigrate system.

ndelangen committed:
Update code/lib/cli-storybook/src/automigrate/fixes/consolidated-imports.ts

Co-authored-by: Valentin Palkovic <valentin@chromatic.com>

ndelangen committed:
Migration: Document dropped support for legacy Storybook packages in 9.0.0

Update MIGRATION.md to detail package consolidation and import changes for Storybook 9.0.0, including:
- List of packages moved to internal paths
- Instructions for package uninstallation and import updates
- Guidance for addon authors on internal package usage

ndelangen committed:
Update MIGRATION.md

Co-authored-by: greptile-apps[bot] <165735046+greptile-apps[bot]@users.noreply.github.com>

valentinpalkovic committed:
Merge branch 'next' into norbert/automigrate-cpc-complaince

valentinpalkovic committed:
Update @storybook/test package mapping

valentinpalkovic committed:
Support dot folders and adjust regex to support sub-paths

ndelangen committed:
Merge branch 'norbert/automigrate-cpc-complaince' into norbert/manual-cpc-migration

ndelangen commented on code/lib/cli-storybook/src/automigrate/helpers/consolidated-packages.ts:
```diff
@@ -0,0 +1,27 @@
+ * object maps the old package name to the new package name.
+ */
+export const consolidatedPackages = {
+  '@storybook/channels': 'storybook/internal/channels',
+  '@storybook/client-logger': 'storybook/internal/client-logger',
+  '@storybook/core-common': 'storybook/internal/common',
+  '@storybook/core-events': 'storybook/internal/core-events',
+  '@storybook/csf-tools': 'storybook/internal/csf-tools',
+  '@storybook/docs-tools': 'storybook/internal/docs-tools',
+  '@storybook/node-logger': 'storybook/internal/node-logger',
+  '@storybook/preview-api': 'storybook/internal/preview-api',
+  '@storybook/router': 'storybook/internal/router',
+  '@storybook/telemetry': 'storybook/internal/telemetry',
+  '@storybook/theming': 'storybook/internal/theming',
+  '@storybook/types': 'storybook/internal/types',
+  '@storybook/manager-api': 'storybook/internal/manager-api',
+  '@storybook/manager': 'storybook/internal/manager',
+  '@storybook/preview': 'storybook/internal/preview',
+  '@storybook/core-server': 'storybook/internal/core-server',
+  '@storybook/builder-manager': 'storybook/internal/builder-manager',
+  '@storybook/components': 'storybook/internal/components',
+  '@storybook/test': 'storybook/test',
```
```suggestion
  '@storybook/channels': 'storybook/internal/channels',
  '@storybook/client-logger': 'storybook/internal/client-logger',
  '@storybook/core-common': 'storybook/internal/common',
  '@storybook/core-events': 'storybook/internal/core-events',
  '@storybook/csf-tools': 'storybook/internal/csf-tools',
  '@storybook/docs-tools': 'storybook/internal/docs-tools',
  '@storybook/node-logger': 'storybook/internal/node-logger',
  '@storybook/preview-api': 'storybook/preview-api',
  '@storybook/router': 'storybook/internal/router',
  '@storybook/telemetry': 'storybook/internal/telemetry',
  '@storybook/theming': 'storybook/theming',
  '@storybook/types': 'storybook/internal/types',
  '@storybook/manager-api': 'storybook/manager-api',
  '@storybook/manager': 'storybook/internal/manager',
  '@storybook/preview': 'storybook/internal/preview',
  '@storybook/core-server': 'storybook/internal/core-server',
  '@storybook/builder-manager': 'storybook/internal/builder-manager',
  '@storybook/components': 'storybook/internal/components',
  '@storybook/test': 'storybook/test',
```

valentinpalkovic committed:
Update MIGRATION.md

Co-authored-by: Norbert de Langen <ndelangen@me.com>

valentinpalkovic committed:
Merge pull request #30766 from storybookjs/norbert/manual-cpc-migration

Migration: Document dropped support legacy packages
